File: crates/vera/src/lexer.rs
//! VERA MVP lexer (Phase 1).

use crate::ast::Span;
use thiserror::Error;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokKind {
    Ident,
    TypeIdent,
    Int,
    Str,
    Kw,
    Op,
    Punct,
    Eof,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokKind,
    pub text: String,
    pub span: Span,
}

#[derive(Debug, Error)]
#[error("{span:?}: {message}")]
pub struct LexError {
    pub message: String,
    pub span: Span,
}

const KEYWORDS: &[&str] = &[
    "fn", "let", "if", "else", "match", "uses", "requires", "ensures", "true", "false", "unit",
    "Int", "Bool", "Str", "Unit", "Console", "List", "Option", "Result", "struct", "enum",
];

pub fn lex(source: &str) -> Result<Vec<Token>, LexError> {
    let chars: Vec<char> = source.chars().collect();
    let n = chars.len();
    let mut i = 0usize;
    let mut line = 1u32;
    let mut col = 1u32;
    let mut out = Vec::new();

    let peek = |i: usize, k: usize| -> Option<char> {
        let j = i + k;
        if j < n {
            Some(chars[j])
        } else {
            None
        }
    };

    while i < n {
        let ch = chars[i];
        // UTF-8 BOM (common on Windows editors / PowerShell Out-File)
        if ch == '\u{FEFF}' {
            i += 1;
            continue;
        }
        if ch == ' ' || ch == '\t' || ch == '\r' {
            i += 1;
            col += 1;
            continue;
        }
        if ch == '\n' {
            i += 1;
            line += 1;
            col = 1;
            continue;
        }
        if ch == '/' && peek(i, 1) == Some('/') {
            while i < n && chars[i] != '\n' {
                i += 1;
            }
            continue;
        }

        let start = Span { line, col };

        if ch.is_ascii_alphabetic() || ch == '_' {
            let mut text = String::new();
            while i < n && (chars[i].is_ascii_alphanumeric() || chars[i] == '_') {
                text.push(chars[i]);
                i += 1;
                col += 1;
            }
            let kind = if KEYWORDS.contains(&text.as_str()) {
                TokKind::Kw
            } else if text.chars().next().is_some_and(|c| c.is_ascii_uppercase()) {
                TokKind::TypeIdent
            } else {
                TokKind::Ident
            };
            out.push(Token {
                kind,
                text,
                span: start,
            });
            continue;
        }

        if ch.is_ascii_digit() {
            let mut text = String::new();
            while i < n && (chars[i].is_ascii_digit() || chars[i] == '_') {
                if chars[i] != '_' {
                    text.push(chars[i]);
                }
                i += 1;
                col += 1;
            }
            out.push(Token {
                kind: TokKind::Int,
                text,
                span: start,
            });
            continue;
        }

        if ch == '"' {
            i += 1;
            col += 1;
            let mut text = String::new();
            while i < n && chars[i] != '"' {
                if chars[i] == '\\' {
                    i += 1;
                    col += 1;
                    if i >= n {
                        return Err(LexError {
                            message: "unterminated string".into(),
                            span: start,
                        });
                    }
                    let esc = chars[i];
                    let mapped = match esc {
                        'n' => '\n',
                        't' => '\t',
                        '"' => '"',
                        '\\' => '\\',
                        _ => {
                            return Err(LexError {
                                message: format!("unknown escape \\{esc}"),
                                span: start,
                            });
                        }
                    };
                    text.push(mapped);
                    i += 1;
                    col += 1;
                } else {
                    text.push(chars[i]);
                    i += 1;
                    col += 1;
                }
            }
            if i >= n || chars[i] != '"' {
                return Err(LexError {
                    message: "unterminated string".into(),
                    span: start,
                });
            }
            i += 1;
            col += 1;
            out.push(Token {
                kind: TokKind::Str,
                text,
                span: start,
            });
            continue;
        }

        let two: String = format!("{}{}", ch, peek(i, 1).unwrap_or('\0'));
        if matches!(
            two.as_str(),
            "==" | "!=" | "<=" | ">=" | "||" | "&&" | "->" | "=>" | "++" | "::"
        ) {
            i += 2;
            col += 2;
            out.push(Token {
                kind: TokKind::Op,
                text: two[..2].to_string(),
                span: start,
            });
            continue;
        }

        if "+-*/%<>=!?".contains(ch) {
            i += 1;
            col += 1;
            // `?ident` with no space → hole token (SPEC §3.2).
            if ch == '?' {
                if let Some(next) = peek(i, 0) {
                    if next.is_ascii_alphabetic() || next == '_' {
                        let mut name = String::new();
                        while i < chars.len()
                            && (chars[i].is_ascii_alphanumeric() || chars[i] == '_')
                        {
                            name.push(chars[i]);
                            i += 1;
                            col += 1;
                        }
                        out.push(Token {
                            kind: TokKind::Op,
                            text: format!("?{name}"),
                            span: start,
                        });
                        continue;
                    }
                }
            }
            out.push(Token {
                kind: TokKind::Op,
                text: ch.to_string(),
                span: start,
            });
            continue;
        }

        if "(){},;:|.[]".contains(ch) {
            i += 1;
            col += 1;
            out.push(Token {
                kind: TokKind::Punct,
                text: ch.to_string(),
                span: start,
            });
            continue;
        }

        return Err(LexError {
            message: format!("unexpected character {ch:?}"),
            span: start,
        });
    }

    out.push(Token {
        kind: TokKind::Eof,
        text: String::new(),
        span: Span { line, col },
    });
    Ok(out)
}

File: compiler/lexer.py
"""VERA MVP lexer (Phase 1)."""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import Iterator

from .ast_nodes import Span


class TokKind(Enum):
    IDENT = auto()
    TYPE_IDENT = auto()
    INT = auto()
    STR = auto()
    KW = auto()
    OP = auto()
    PUNCT = auto()
    EOF = auto()


KEYWORDS = {
    "fn",
    "let",
    "if",
    "else",
    "uses",
    "requires",
    "ensures",
    "true",
    "false",
    "unit",
    "Int",
    "Bool",
    "Str",
    "Unit",
    "Console",
    "List",
    "Option",
    "Result",
}


@dataclass(frozen=True)
class Token:
    kind: TokKind
    text: str
    span: Span


class LexError(Exception):
    def __init__(self, message: str, span: Span) -> None:
        super().__init__(f"{span.line}:{span.col}: {message}")
        self.span = span


def lex(source: str) -> list[Token]:
    i = 0
    line = 1
    col = 1
    n = len(source)
    out: list[Token] = []

    def peek(k: int = 0) -> str:
        j = i + k
        return source[j] if j < n else ""

    def advance() -> str:
        nonlocal i, line, col
        ch = source[i]
        i += 1
        if ch == "\n":
            line += 1
            col = 1
        else:
            col += 1
        return ch

    def span_here() -> Span:
        return Span(line, col)

    while i < n:
        ch = peek()
        if ch in " \t\r\n":
            advance()
            continue
        if ch == "/" and peek(1) == "/":
            while i < n and peek() != "\n":
                advance()
            continue

        start = span_here()

        if ch.isalpha() or ch == "_":
            text = ""
            while i < n and (peek().isalnum() or peek() == "_"):
                text += advance()
            if text in KEYWORDS:
                # Built-in type names are keywords; value keywords too.
                if text[0].isupper() and text not in {"true", "false", "unit"}:
                    out.append(Token(TokKind.KW, text, start))
                elif text in {"true", "false", "unit", "fn", "let", "if", "else", "uses", "requires", "ensures"}:
                    out.append(Token(TokKind.KW, text, start))
                else:
                    out.append(Token(TokKind.KW, text, start))
            elif text[0].isupper():
                out.append(Token(TokKind.TYPE_IDENT, text, start))
            else:
                out.append(Token(TokKind.IDENT, text, start))
            continue

        if ch.isdigit():
            text = ""
            while i < n and (peek().isdigit() or peek() == "_"):
                text += advance()
            out.append(Token(TokKind.INT, text.replace("_", ""), start))
            continue

        if ch == '"':
            advance()
            text = ""
            while i < n and peek() != '"':
                if peek() == "\\":
                    advance()
                    esc = advance()
                    mapping = {"n": "\n", "t": "\t", '"': '"', "\\": "\\"}
                    if esc not in mapping:
                        raise LexError(f"unknown escape \\{esc}", start)
                    text += mapping[esc]
                else:
                    text += advance()
            if peek() != '"':
                raise LexError("unterminated string", start)
            advance()
            out.append(Token(TokKind.STR, text, start))
            continue

        two = ch + peek(1)
        if two in {"==", "!=", "<=", ">=", "||", "&&", "->", "++"}:
            advance()
            advance()
            out.append(Token(TokKind.OP, two, start))
            continue

        if ch in "+-*/%<>=!":
            out.append(Token(TokKind.OP, advance(), start))
            continue

        if ch in "(){},;:|.":
            out.append(Token(TokKind.PUNCT, advance(), start))
            continue

        raise LexError(f"unexpected character {ch!r}", start)

    out.append(Token(TokKind.EOF, "", Span(line, col)))
    return out
